### PR TITLE
[feature][METEOR-590] Implement reading raw coordinates of TFS stage

### PR DESCRIPTION
**Pull request by K4rishma**
- read raw coordinates instead of linked coordinates
- correct for offset in x and y such that the coordinated read by TFS and odemis is same in raw coordinate system
- This change requires v1.12.0 of xtadapter on support PC

### CODE REVIEW

File: src/odemis/driver/xt_client.py [hunk 1]
```diff
@@ -358,6 +358,16 @@ def check_and_transfer_latest_package(self) -> None:
...
                     logging.warning("{} is a bad file in {} not transferring latest package.".format(ret, package.path))
+
+                # This version supports reading raw stage coordinates
+                if current_version is not None and bitness is not None:
+                    if current_version <= "1.12.0":
+                        raise ValueError("Required Delmic XT Adapter version shoulb be >= 1.12.0",
+                                         "How to update?\n\n1. Full stop Odemis and close Delmic XT Adapter.\n"
+                                         "2. Uninstall the old Delmic XT Adapter and install the correct version.")
+                else:
+                    logging.warning(
```
Comment (K4rishma) [inline src/odemis/driver/xt_client.py:369]: Should this be an error ? How about if the Delmix XT adaptor is in development version?

File: src/odemis/driver/xt_client.py [hunk 2]
```diff
@@ -2170,7 +2196,7 @@ def __init__(self, name: str, role: str, parent: model.HwComponent, rng: dict =
             "rz": model.Axis(unit=stage_info["unit"]["r"], range=rng["rz"]),
         }
 
-        model.Actuator.__init__(self, name, role, parent=parent, axes=axes_def,
+        model.Actuator.__init__(self, name, role, parent=self.parent, axes=axes_def,
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:2199]: This function takes care of setting `self.parent` as `parent`. So no need to do it early explicitly.

File: src/odemis/driver/xt_client.py [hunk 3]
```diff
@@ -2147,10 +2166,17 @@ class Stage(model.Actuator):
     moving the TFS stage and updating the position.
     """
 
-    def __init__(self, name: str, role: str, parent: model.HwComponent, rng: dict = None, **kwargs) -> None:
+    def __init__(self, name: str, role: str, parent: model.HwComponent, rng: dict = None, raw: bool = False,
+                 **kwargs) -> None:
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:2170]: This needs a docstring. At least it should document the new "raw" argument. It'd be a good opportunity to document rng too.

File: src/odemis/driver/xt_client.py [hunk 4]
```diff
@@ -2147,10 +2166,17 @@ class Stage(model.Actuator):
     moving the TFS stage and updating the position.
     """
 
-    def __init__(self, name: str, role: str, parent: model.HwComponent, rng: dict = None, **kwargs) -> None:
+    def __init__(self, name: str, role: str, parent: model.HwComponent, rng: dict = None, raw: bool = False,
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:2169]: "raw" a little bit generic word. I think that when encountering it in a microscope file, a lot of people wouldn't understand what it refers too. You can of course count on comments, but I'd suggest to name it `raw_coordinates` to make it clearer immediately.

File: src/odemis/driver/xt_client.py [hunk 5]
```diff
@@ -2147,10 +2166,17 @@ class Stage(model.Actuator):
...
+                 **kwargs) -> None:
+        self.raw = raw
+        self.parent = parent
+        # define axes which needs offset correction such that TFS stage coordinates
+        # are equal to stage coordinates displayed through Odemis
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:2174]: Please explain more this comment. At least you should explain that 1) it's related to the raw coordinates, and 2) it'll be updated later in the init.

File: src/odemis/driver/xt_client.py [hunk 6]
```diff
@@ -2147,10 +2166,17 @@ class Stage(model.Actuator):
...
+        self.offset = {"x": 0, "y": 0}
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:2175]: Let's a use a more specific name like `_raw_offset`.

File: src/odemis/driver/xt_client.py [hunk 7]
```diff
@@ -2147,10 +2166,17 @@ class Stage(model.Actuator):
     moving the TFS stage and updating the position.
     """
 
-    def __init__(self, name: str, role: str, parent: model.HwComponent, rng: dict = None, **kwargs) -> None:
+    def __init__(self, name: str, role: str, parent: model.HwComponent, rng: dict = None, raw: bool = False,
+                 **kwargs) -> None:
+        self.raw = raw
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:2171]: It's private -> `_raw` (or `_raw_coordinates`)

File: src/odemis/driver/xt_client.py [hunk 8]
```diff
@@ -2179,16 +2205,42 @@ def __init__(self, name: str, role: str, parent: model.HwComponent, rng: dict =
...
+            self.parent.set_raw_coordinate_system(self.raw)
+            pos = self._getPosition()
+            logging.debug(f"Raw stage coordinates {pos}")
+            self.parent.set_raw_coordinate_system(False)
+            pos_after = self._getPosition()
+            logging.debug(f"Linked stage coordinates {pos_after}")
+            for axis in self.offset.keys():
+                self.offset[axis] = pos_after[axis] - pos[axis]
+            # the offset should only be in linear axes, it is not expected in rotational axes
+            assert all(pos[axis] == pos_after[axis] for axis in ["rx", "rz"])
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:2228]: It's interesting to check, but let's not immediately completely fail in this case. Just log a warning.

File: src/odemis/driver/xt_client.py [hunk 9]
```diff
@@ -2179,16 +2205,42 @@ def __init__(self, name: str, role: str, parent: model.HwComponent, rng: dict =
                                                 readonly=True)
         self._updatePosition()
 
+        # update the offset if raw stage coordinates are read
+        self._get_offset_raw_coordinate_system()
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:2209]: This function currently does actually 2 things: measure the offset, and change the raw coordinates. The name implies it just does one (and return the value, which it doesn't do).
I'd suggest to call it like `self._switch_coordinate_system(self._raw_coordinates)`. (it takes 1 argument)

File: src/odemis/driver/xt_client.py [hunk 10]
```diff
@@ -2179,16 +2205,42 @@ def __init__(self, name: str, role: str, parent: model.HwComponent, rng: dict =
...
+
         # Refresh regularly the position
         self._pos_poll = util.RepeatingTimer(5, self._refreshPosition, "Stage position polling")
         self._pos_poll.start()
 
+    def _get_offset_raw_coordinate_system(self) -> None:
+        """Calculate the offset in linear stage axes x and y when the stage coordinates are read in raw coordinates such
+         that it is equal to x and y when stage coordinates are read in linked coordinates."""
+        if self.raw:
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:2218]: If not raw, it should also explicitly change the raw coordinate system, so that we are certain it's in the coordinate system we expect.
If it fails because the function doesn't exist, it's fine (as that's typically the standard value).

File: src/odemis/driver/xt_client.py [hunk 11]
```diff
@@ -2179,16 +2205,42 @@ def __init__(self, name: str, role: str, parent: model.HwComponent, rng: dict =
...
+            self.parent.set_raw_coordinate_system(self.raw)
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:2219]: Instead of changing 3 times, start with the "non raw coordinates" system, and then switch to the raw version.

File: src/odemis/driver/xt_client.py [hunk 12]
```diff
@@ -2179,16 +2205,42 @@ def __init__(self, name: str, role: str, parent: model.HwComponent, rng: dict =
...
         """
         update the position VA
         """
         old_pos = self.position.value
         pos = self._getPosition()
+        if self.raw:
+            # correct for the offset such that the stage coordinates displayed in
+            # TFS software is the same as Odemis
+            pos["x"] = pos["x"] + self.offset["x"]
+            pos["y"] = pos["y"] + self.offset["y"]
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:2243]:
```suggestion
            pos["x"] += self.offset["x"]
            pos["y"] += self.offset["y"]
```

File: src/odemis/driver/xt_client.py [hunk 13]
```diff
@@ -2235,6 +2287,13 @@ def _moveTo(self, future: futures.Future, pos: Dict[str, float], rel: bool = Fal
                     pos["t"] = pos.pop("rx")
                 if "rz" in pos.keys():
                     pos["r"] = pos.pop("rz")
+
+                # Correct for offset compensation in absolute movements
+                if self.raw and not rel:
+                    if "x" in pos.keys():
+                        pos["x"] = pos["x"] - self.offset["x"]
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:2294]:
```suggestion
                        pos["x"] -= self.offset["x"]
```
(same for y)

File: src/odemis/driver/xt_client.py [hunk 14]
```diff
@@ -2179,16 +2205,42 @@ def __init__(self, name: str, role: str, parent: model.HwComponent, rng: dict =
...
+            pos = self._getPosition()
+            logging.debug(f"Raw stage coordinates {pos}")
+            self.parent.set_raw_coordinate_system(False)
+            pos_after = self._getPosition()
+            logging.debug(f"Linked stage coordinates {pos_after}")
+            for axis in self.offset.keys():
+                self.offset[axis] = pos_after[axis] - pos[axis]
+            # the offset should only be in linear axes, it is not expected in rotational axes
+            assert all(pos[axis] == pos_after[axis] for axis in ["rx", "rz"])
+            logging.debug(f"The offset values in x and y are {self.offset} when stage is in the raw coordinate system")
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:2229]: Please only put 1 logging message for this function. It should say the non-raw & raw values, and conclude on the offset, as a single sentence.

File: src/odemis/driver/xt_client.py [hunk 15]
```diff
@@ -376,6 +386,15 @@ def move_stage(self, position: Dict[str, float], rel: bool = False) -> None:
             self.server._pyroClaimOwnership()
             self.server.move_stage(position, rel)
 
+    def set_raw_coordinate_system(self, raw: bool = True):
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:389]: Even if you don't use it now, can you add the `get_raw_coordinate_system()` function so that we know it exists?

File: src/odemis/driver/xt_client.py [hunk 16]
```diff
@@ -358,6 +358,16 @@ def check_and_transfer_latest_package(self) -> None:
                     update.show()
                 else:
                     logging.warning("{} is a bad file in {} not transferring latest package.".format(ret, package.path))
+
+                # This version supports reading raw stage coordinates
+                if current_version is not None and bitness is not None:
+                    if current_version <= "1.12.0":
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:364]: FYI, you cannot compare versions with string comparison. For instance "9" > "10"! There are dedicated functions for that.

File: src/odemis/driver/xt_client.py [hunk 17]
```diff
@@ -358,6 +358,16 @@ def check_and_transfer_latest_package(self) -> None:
                     update.show()
                 else:
                     logging.warning("{} is a bad file in {} not transferring latest package.".format(ret, package.path))
+
+                # This version supports reading raw stage coordinates
+                if current_version is not None and bitness is not None:
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:363]: Don't check explicitly for the version. Instead, I would suggest that if (and *only if*) the raw coordinates are requested, and the call to `set_raw_coordinate_system()`  fails *because the method doesn't exist*, then you raise a ValueError (or maybe, more appropriately, a NotImplementedError).

So, in particular, if the user asks raw_coordinates = False, and calling the `set_raw_coordinate_system(False)` fails, it should continue business as usual.

**Review by pieleric — CHANGES_REQUESTED** [review]

**Comment by K4rishma** [comment]
Ready for review

File: src/odemis/driver/xt_client.py [hunk 18]
```diff
@@ -376,6 +377,24 @@ def move_stage(self, position: Dict[str, float], rel: bool = False) -> None:
...
+    def set_raw_coordinate_system(self, raw_coordinates:bool = True):
+        """
+        Read raw z coordinate or linked z coordinate when requesting stage coordinates
+        :param raw_coordinates: (bool) If True, request raw z stage coordinates, otherwise request linked Z coordiantes
+        """
+        try:
+            with self._proxy_access:
+                self.server._pyroClaimOwnership()
+                self.server.set_raw_coordinate_system(raw_coordinates)
+        except AttributeError:
```
Comment (tepals) [inline src/odemis/driver/xt_client.py:389]: are you sure this will raise an `AttributeError`? I thought Pyro5 translated everything into an `OSError`

File: src/odemis/driver/xt_client.py [hunk 19]
```diff
@@ -376,6 +377,24 @@ def move_stage(self, position: Dict[str, float], rel: bool = False) -> None:
...
+            if raw_coordinates:
+                raise NotImplementedError("Ensure the version of Delmic XT Adapter is >= 1.12.0")
+            # If non-raw coordinates are requested, proceed with the current coordinate system
```
Comment (tepals) [inline src/odemis/driver/xt_client.py:392]:
```suggestion
                raise NotImplementedError("Ensure the version of Delmic XT Adapter is >= 1.12.0")
            # Do not raise an error if non-raw coordinates are requested, because non-raw is the default in old versions of the xtadapter
```

File: src/odemis/driver/xt_client.py [hunk 20]
```diff
@@ -376,6 +377,24 @@ def move_stage(self, position: Dict[str, float], rel: bool = False) -> None:
...
+
+    def get_raw_stage_position(self):
+        """Read the stage position in raw coordinate system"""
+        pass
```
Comment (tepals) [inline src/odemis/driver/xt_client.py:396]: why is this here?

**Review by tepals — CHANGES_REQUESTED** [review]

File: src/odemis/driver/xt_client.py (hunk 15 above)
Comment (K4rishma) [reply src/odemis/driver/xt_client.py]: @tepals the reason of adding this function is explained here

Comment (tepals) [reply src/odemis/driver/xt_client.py]: I think what Éric meant (but correct me if I'm wrong) is that since you have the function on the xtadapter side, you might as well implement it here. Now you just have it implemented as an empty function, while you could also implement the working call to the xtadapter. So not just do `pass` but actually call the function.

File: src/odemis/driver/xt_client.py (hunk 18 above)
Comment (K4rishma) [reply src/odemis/driver/xt_client.py]: I am not sure about it either, I have added it like below. In SEM class, I see mostly timeout errors and I see OS errors in Detector class
`        except OSError:
            if raw_coordinates:
                raise NotImplementedError("Ensure the version of Delmic XT Adapter is >= 1.12.0")`

File: src/odemis/driver/xt_client.py (hunk 20 above)
Comment (K4rishma) [reply src/odemis/driver/xt_client.py:397]: based on Eric's remark before I added this function

File: src/odemis/driver/xt_client.py (hunk 15 above)
Comment (pieleric) [reply src/odemis/driver/xt_client.py]: Yes, I confirm this interpretation from Thera :-)
You have to actually implement `get_raw_coordinate_system()`: call the xtadapter's function and return the returned value)

File: src/odemis/driver/xt_client.py [hunk 21]
```diff
@@ -376,6 +377,25 @@ def move_stage(self, position: Dict[str, float], rel: bool = False) -> None:
...
+        """
+        try:
+            with self._proxy_access:
+                self.server._pyroClaimOwnership()
+                self.server.set_raw_coordinate_system(raw_coordinates)
+        except OSError:
+            if raw_coordinates:
+                raise NotImplementedError("Ensure the version of Delmic XT Adapter is >= 1.12.0")
+            # Do not raise an error if non-raw coordinates are requested, because non-raw is the default in old
+            # versions of the xtadapter
```
Comment (pieleric) [inline src/odemis/driver/xt_client.py:393]: This shouldn't be hidden for every call. Only at init, when the `raw_coordinates` argument is `False` (which is the default). => move this `try`/`except` to line 2246.

**Review by pieleric — CHANGES_REQUESTED** [review]

**Review by pieleric — APPROVED** [review]

**Review by patrickcleeve2 — APPROVED** [review]
